### PR TITLE
be careful Currying

### PR DESCRIPTION
**Pull request by jasonmg**

### CODE REVIEW

**Comment by demobox** [comment]
Many thanks for submitting this, @jasonmg! There is some similarity with ["Count Me Now, Count Me Later"](http://scalapuzzlers.com/#pzzlr-022) but I think the varargs here make this sufficiently different.

I'll try to take a look and respond with comments today or tomorrow. Thanks again!

File: puzzlers/pzzlr-067.html [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  <p>
+    the line: optList foreach queue.enqueue, actually we can change to 
+    val enqueueFun = queue.enqueue _     // this is Currying
+    optList foreach enqueueFun
+    the enqueueFun type is (Seq[Any]) => Unit
+
+    and the line: optList.foreach{s => queue1.enqueue(s)}, we can change to
+    val enqueueFun1 = queue.enqueue(_)  // this is normal function, i.e. a => queue.enqueue(a)
+    optList foreach enqueueFun1
+    the enqueueFun1 type is (Any) => Unit
```
Comment (demobox) [inline puzzlers/pzzlr-067.html]: I am not sure if this explanation is accurate, since the type of both the anonymous function and the partial application are the same, according to my REPL (2.11.6):

```
// anonymous function
scala> queue.enqueue(_)
res65: Seq[Any] => Unit = <function1>

// partial application
scala> queue.enqueue _
res67: Seq[Any] => Unit = <function1>
```

From what I can see, the different result is caused by the fact that the compiler somehow is able to deduce in the first case that the argument is still a varargs argument, whereas in the second case that information is lost. Here's how the calls compile:

```
        private[this] val anonymousFun: Seq[Any] => Unit = ((x$1: Any*) => $lin
e231.$read.$iw.$iw.queue1.enqueue(x$1));
...
        private[this] val partialFun: Seq[Any] => Unit = {
          ((elems: Seq[Any]) => $line230.$read.$iw.$iw.queue.enqueue((elems: _*)
))
        };
```

We can achieve the same result in both cases if we write the partial application like this:

```
// urgh!
optList foreach { queue1.enqueue((_: Seq[Any]): _*) }

// slightly better
optList foreach { list => queue1.enqueue(list: _*) }
```

I'll have a look in SLS to see what it has to say about this...

Comment (demobox) [inline puzzlers/pzzlr-067.html]: > I'll have a look in SLS to see what it has to say about this...

Hm, not much, it seems. 6.23 and 6.26.5 describe placeholder syntax for anonymous functions and eta expansion, respectively, but they don't give any details on how to treat a varargs parameter.

Specifically, the [eta-expansion example](http://www.scala-lang.org/files/archive/spec/2.11/06-expressions.html#eta-expansion) does **not** show the last argument as being treated as a `varargs` argument if that's applicable, although that is what the compiler seems to be doing.

Sounds like a good question for the mailing list... ;-)

**Comment by som-snytt** [comment]
You might remember the behavior was changed/fixed. [SI-4176](https://issues.scala-lang.org/browse/SI-4176)
```
$ scala210 -Yeta-expand-keeps-star
Welcome to Scala version 2.10.5 (OpenJDK 64-Bit Server VM, Java 1.7.0_95).
Type in expressions to have them evaluated.
Type :help for more information.

scala> import scala.collection.mutable
import scala.collection.mutable

scala> val queue = new mutable.Queue[Any]()
queue: scala.collection.mutable.Queue[Any] = Queue()

scala> val optList = Some(List(1,2,3))
optList: Some[List[Int]] = Some(List(1, 2, 3))

scala> optList foreach queue.enqueue
<console>:11: error: type mismatch;
 found   : Any* => Unit
 required: List[Int] => ?
              optList foreach queue.enqueue
                                    ^
```
It makes sense that the `A*` receives the "pass-thru" treatment on eta expansion.

This issue of inference difference between eta expansion and anon fun comes up regularly, but it's still a fun example.

**Comment by demobox** [comment]
> You might remember the behavior was changed/fixed. SI-4176

I didn't remember that one, actually - thanks for bringing it up, @som-snytt!

Irrespective of whether the argument type of the resulting function is `A*` or `Seq[A]`, the question for me is still whether there's any explanation why the argument is _passed on_ as a vararg in eta expansion, i.e. why eta expansion results in `(a: Seq[A]) => foo(a: _*)` rather than `(a: Seq[A]) => foo(a)`, which is what the anonymous function becomes.

6.26.5 doesn't seem to have anything specific to say about this - in fact, a quick reading of:

```
The result of eta-conversion is then:

{ 
  val x1 = e1;
  ...
  val xm = em;
  (y1:T1,…,yn:Tn) => e′(y1,…,yn)
}
```
would seem to imply that that we would get `(a: Seq[A]) => foo(a)` (with `yn` being `a` and `Tn` being `Seq[A]` here). I can't see anything obvious that would state that, "if the last parameter to e' is a repeated parameter and Tn conforms to Seq[Whatever], then the invocation should be `e′(y1,…,yn: _*)`" or so.

Curious to hear your thoughts on that!

**Comment by som-snytt** [comment]
In the function case, the param type is inferred from the expected type, so it's `List[Int]`.

For the eta expansion, e' is just e with some subexpressions pre-computed.

So the old behavior, `(a: A*) => f(a: _*)` makes most sense. It still makes sense taking `A*` as `Seq[A]`. In the body, it's always `Seq[A]` anyway.

The other change which did get a mention in the spec was that by-name args are not computed during eta expansion, but are passed through. The theme is, "Don't molest my application."

I think the `Any` kind of spoils the puzzler as such. Maybe there should be a separate category for "cautionary type tales". Or maybe one category, no pun intended, for Typelevel puzzlers, and another for Wild West puzzlers. But it's still educational about Any and intersections of features.

The other angle is that they just closed [the ticket about eta expansion vs empty application](https://issues.scala-lang.org/browse/SI-7187) because they deprecated eta expansion of nullary f. However, they didn't actually switch the order of conversions. In this case, `f()` works for `f(A*)`, but I guess they don't offer empty application for that case.

**Comment by demobox** [comment]
> The other angle is that they just closed the ticket about eta expansion vs empty application because they deprecated eta expansion of nullary f

I was following that one, too - seems like a good step to me, even if it ends up nuking ["Unary Quandary"](http://scalapuzzlers.com/#pzzlr-060). As regards this code sample, I agree that it would be nicer without the `Any` fudge - we can tighten that to `AnyRef`, I think, but that's about it.

My comment/question about _why_ this happens wasn't clear, though - sorry. Let me try again: from my reading, if I follow the instructions in 6.26.5 literally, the eta expansion of `queue.enqueue` in `optList foreach queue.enqueue` seems to come out as:

```
{
  // no vals as nothing to bind
  (y1: Seq[Any]) => queue.enqueue(y1)
}
```

which is **not** what is happening, since we actually get the equivalent of `{ (y1: Seq[Any]) => queue.enqueue(y1: _*) }`.

Where in the spec is `_*` coming from? Or is it a case of "common sense clarifying/refining the spec"?

**Comment by som-snytt** [comment]
Yeah, your question was clear. And you're right the spec isn't clear.

If I want to invoke `f(A*)`, then the function body must invoke `f(A*)`, not `f(Seq[A])`. If f is overloaded, don't call the wrong one. (I didn't try if that could happen.)

**Comment by demobox** [comment]
> Yeah, your question was clear. And you're right the spec isn't clear.

Thanks for, er, "clarifying". Have a good Sunday!

**Comment by demobox** [comment]
@jasonmg @som-snytt Still struggling with this one because it really seems more like an "anti-puzzler", i.e. the code behaves frustratingly in line with expectations even though (as discussed above) it's not clear how or why it does.

The first challenge I have is that the `_*` type annotation is sufficiently common that `Some(List(...)) foreach { list=> enqueue(list) }` simply "looks wrong", i.e. is not something one would be likely to see without expecting a bug of some kind. So the behaviour in that case is not all that surprising - I suspect many would correctly guess that the list is not, in fact, expanded to a varargs argument.

The second challenge is that the eta expansion also behaves according to most people's expectation, I think - I would suspect that people would indeed assume that `Some(List(...)) foreach enqueue` "does the right thing" with respect to varargs, even though there's much more of a mystery here than might be assumed on the face of it.

I think the best hope here would be for the first example to make readers consider how the second example should behave, but I still feel that, in a code sample like the following, the correct answer would still be what most would consider to be "the obvious one":

```
def assembleAndPrint(fragments: Any*) = println(fragments.mkString)

val channel = Seq(Seq("An","swe", 'r'), Seq(' ', "is ", 42)) // different types here

channel foreach { assembleAndPrint(_) }
channel foreach assembleAndPrint // or ... 'assembleAndPrint _', but seems more idiomatic without the trailing '_'?
```

Curious to hear any thoughts/comments/suggestions!

File: puzzlers/pzzlr-067.html (hunk 1 above)
Comment (retronym) [inline puzzlers/pzzlr-067.html]: I'm pretty sure there is a compiler bug in the mix.

```
scala> List.apply[Int](_) //print
   ((x$1: Int*) => scala.collection.immutable.List.apply[scala.Int](x$1)) // : Int* => List[Int]

scala> List.apply[Int] _ //print
   {
  ((xs: Seq[Int]) => scala.collection.immutable.List.apply[scala.Int](((xs): _*)))
} // : Seq[Int] => List[Int]
```

We should not have an expression of type `Int* => ...`.

I notice the compiler has:

```scala
      // dropIllegalStarTypes is a localized fix for SI-6897. We should probably
      // integrate that transformation at a lower level in master, but lubs are
      // the likely and maybe only spot they escape, so fixing here for 2.10.1.
      existentialAbstraction(tparams, dropIllegalStarTypes(lubType))
```

Comment (demobox) [inline puzzlers/pzzlr-067.html]: > We should not have an expression of type Int* => ...

@retronym Wow, that does indeed look funky ;-) Just to confirm that I'm seeing the same on 2.11.6, this is what you're referring to (using `-Xprint:typer`):
```
scala> List.apply[Any](_)
...
        private[this] val res14: Seq[Any] => List[Any] = ((x$1: Any*) => immutab
le.this.List.apply[Any](x$1));
```
?

The _declared_ type seems to be correct, though - the result does not appear to behave as though it has a varargs parameter, although I wouldn't be surprised if you had a trick or two up your sleeve to get at that `Any*` ;-)

```
scala> res14(Seq(1, "2", 3))
res15: List[Any] = List(List(1, 2, 3))

scala> res14(1, "2", 3)
<console>:9: error: too many arguments for method apply: (v1: Seq[Any])List[Any]
 in trait Function1
              res14(1, "2", 3)
                   ^
```

Comment (som-snytt) [inline puzzlers/pzzlr-067.html]: @demobox That's using Jason's ace in the hole, or is it up his sleeve: `//print[tab]` for which I guess you need 2.11.8 or later.

Your version works with Any, it's the Int example that errors. Obviously, the correct syntax is:
```
List.apply[(Int,)](_)
```

**Comment by som-snytt** [comment]
Good one! Printing is much better motivation for taking Any. I would publish the puzzler quick, before Jason breaks it. (Or, from his demented point of view, "fixes" it.)

To me, seeing `f(_)` and `f _` side-by-side makes the difference in result striking.

Here's another look. You almost never take these suggestions, but maybe it will help you arrive at an interesting narrative.

```
$ scala
Welcome to Scala 2.11.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_92).
Type in expressions for evaluation. Or try :help.

scala> implicit class tester(val sc: StringContext) {
     | val f = sc.s _    // or val f: Seq[Any] => String = sc.s
     | def x(args: Any*) = f(args)
     | }
defined class tester

scala> implicit class tester2(val sc: StringContext) {
     | val f = sc.s(_: Seq[Any])   // or just Any, or val f: Seq[Any] => String = sc.s(_)
     | def y(args: Any*) = f(args)
     | }
defined class tester2

scala> val a = "Bob" ; val b = "Betty"
a: String = Bob
b: String = Betty

scala> x"Hi, $a. Bye, $b."
res0: String = Hi, Bob. Bye, Betty.

scala> y"Hi, $a. Bye, $b."
java.lang.IllegalArgumentException: wrong number of arguments (1) for interpolated string with 3 parts
  at scala.StringContext.checkLengths(StringContext.scala:65)
  at scala.StringContext.standardInterpolator(StringContext.scala:120)
  at scala.StringContext.s(StringContext.scala:95)
  at tester2$$anonfun$1.apply(<console>:13)
  at tester2$$anonfun$1.apply(<console>:13)
  at tester2.y(<console>:14)
  ... 32 elided

```
I like the version with `val f: Seq[Any] => String` because usually we say that if the types align, the program will just work. This shape looks more specific than Any, so it's tricky.

File: puzzlers/pzzlr-067.html (hunk 1 above)
Comment (demobox) [inline puzzlers/pzzlr-067.html]: > //print[tab] for which I guess you need 2.11.8

I suspect I'm due for a quick REPL upgrade ;-) Looks like I'm getting different behaviour in 2.11.6, though, unless @retronym's first example doesn't actually compile:

```
scala> List.apply[Int](_)
<console>:8: error: type mismatch;
 found   : Seq[Int]
 required: Int
              List.apply[Int](_)
                              ^

scala> List.apply[Int](_: Int)
res2: Int => List[Int] = <function1>

scala> List.apply[Int]((_: Seq[Int]): _*)
res5: Seq[Int] => List[Int] = <function1>
```

`-Xprint:typer` on the latter seems to indicate that the declared type does not "leak" a vararg parameter as in Jason's `// : Int* => List[Int]` snippet. Here, that's likely because of the type ascription, but in the `Any` case above the declared type was still `Seq[Any] => ...`, rather than `Any* => ...`:

```
        private[this] val res6: Seq[Int] => List[Int] = ((x$1: Seq[Int]) => immutable.this.List.apply[Int](((x$1: Seq[Int]): _*)));
```

**Comment by demobox** [comment]
> Here's another look

String contexts! Excellent idea! I think that's definitely one we'll want to run with... ;-)

Thanks!

**Comment by demobox** [comment]
@jasonmg @som-snytt How about this? Essentially, just wrapping up the string interpolation idea in a bit more indirection to hopefully motivate why you might want to use this (actually, context-specific rendering seems quite a nice idea...):

```
import language.implicitConversions

trait RenderContext extends (StringContext => (Seq[Any] => String))

implicit class MyCtxHelper(val sc: StringContext) extends AnyVal {
  def my(args: Any*)(implicit rc: RenderContext) = rc(sc).apply(args)
} 
 
val name = "Scala"
val greeting = "I like your dots!"

{
  implicit val useS = 
    new RenderContext { def apply(sc: StringContext) = sc.s _ }
  println(my"Hello $name,\n$greeting")
}

{
  implicit val useRaw = 
    new RenderContext { def apply(sc: StringContext) = sc.raw(_) }
  println(my"Hello $name,\n$greeting")
}
```

With the pleasantly puzzling outcome:

```
scala> {
     |   implicit val useS =
     |     new RenderContext { def apply(sc: StringContext) = sc.s _ }
     |   println(my"Hello $name,\n$greeting")
     | }
Hello Scala,
I like your dots!

scala>

scala> {
     |   implicit val useRaw =
     |     new RenderContext { def apply(sc: StringContext) = sc.raw(_) }
     |   println(my"Hello $name,\n$greeting")
     | }
java.lang.IllegalArgumentException: wrong number of arguments (1) for interpolated string with 3 parts
  at scala.StringContext.checkLengths(StringContext.scala:65)
  at scala.StringContext.standardInterpolator(StringContext.scala:120)
  at scala.StringContext.raw(StringContext.scala:117)
  at $anon$1$$anonfun$apply$1.apply(<console>:16)
  at $anon$1$$anonfun$apply$1.apply(<console>:16)
  at MyCtxHelper$.my$extension(<console>:10)
  ... 41 elided
```

The only thing I'm a bit concerned about is whether readers will get distracted too much by thinking of what `raw` would do. One option would be to have a string with no special characters, or to use `s` twice...but then the bit you're supposed to focus one (i.e. eta expansion vs. partial application of the function) is _very_ clear.

**Comment by jasonmg** [comment]
@demobox @som-snytt  you gays are amazing, i didn't dig so deeply about this,  *StringContext* example is great and very straightforward(i thought both using *s* instead of using *raw* at latter is better), i would like leave it to you gays to decide where it gonna be puzzler or not?

**Comment by som-snytt** [comment]
For the record, we share a first name but are not actually a couple.

**Comment by demobox** [comment]
> For the record, we share a first name but are not actually a couple.

The "Andrew & Andrew" show? ;-)

**Comment by SethTisue** [comment]
A and P

**Comment by demobox** [comment]
@SethTisue: I hope the "Store Closing" isn't intended as a reference... ;-)

**Comment by jasonmg** [comment]
OH MY,  APOLOGIES for my stupid typo, i even don't know how to say thank you for your such elaborate analysis, please don't get me wrong, i just start mainly use scala 6month ago, still in learning stage, AGAIN,   APOLOGIES AND THANK YOU.

**Comment by demobox** [comment]
> OH MY, APOLOGIES for my stupid typo

From my perspective, there's no need to apologize at all - typos happen to all of us (I can't imagine the kinds of things I would be saying if I tried to write in Chinese!), and I very much had the feeling that it was unintentional.

It's great to see someone taking the effort to submit a puzzler, especially someone relatively new to the language. Keep it going!

**Comment by demobox** [comment]
Discussion to be continued over at #145? Thanks again for submitting, @jasonmg!
